include/deal.II/base/utilities.h

```diff
@@ -347,12 +347,18 @@ namespace Utilities
 
   /**
    * A replacement for <code>std::pow</code> that allows compile-time
-   * calculations for constant expression arguments.
+   * calculations for constant expression arguments. The exponent @p iexp
+   * must not be negative.
    */
   constexpr unsigned int
-  pow(const unsigned int base, const unsigned int iexp)
+  pow(const unsigned int base, const int iexp)
   {
-    return iexp == 0 ? 1 : base * dealii::Utilities::pow(base, iexp - 1);
+#ifdef DEAL_II_WITH_CXX17
+    AssertThrow(iexp >= 0, ExcMessage("The exponent must not be negative!"));
```

Comment:
Why does this statement require C++17? Because it would not be allowed inside a `constexpr` function otherwise?

Relatedly, why use `AssertThrow` instead of `Assert`?

Reply:
This function can be called with input data that can change from run to run. Hence, I would argue that checking in debug mode is not sufficient.

Reply:
Hm, that's no more true than almost any other function in deal.II as well. Is the function called from paths that, for example, directly read from input files?

Reply:
To be honest, C++14 should be sufficient to call `throw` in a `constexpr` function (without doing that in the return statement). However, some compilers claiming to understand C++14 (e.g.`gcc-4.9.0`-`gcc-5.5`) can't deal with that. I didn't want to disable `C++14` support just because of this issue. All the compiler supporting `C++17` I could find on `godbolt`, don't have this deficiency. Of course, I can still introduce a compiler check instead if you prefer that.

Reply:
Is the issue that we cannot use `Assert` in `constexpr` code?

Reply:
`std::abort` can't be used with up to (including) `gcc-4.9.4` with `std=c++14` within a `constexpr` function.

Reply:
> Hm, that's no more true than almost any other function in deal.II as well. Is the function called from paths that, for example, directly read from input files?

No, currently we don't do that within the library. Still, I consider this a quite high-level function.

In the end, I am also happy with using `Assert` instead. Then, I see however no chance to test for compiler support. Since at least the `DEAL_II_HAVE_CXX14_CONSTEXPR_STDMAXMIN` fails for `gcc-4.9.0`-`gcc-4.9.4`, we should be fine requiring `C++14` though. 

Since both of you seem to feel more strongly about this than me, let me change that and go with the less restrictive C++ standard support check.

Reply:
I'm happy with the current `#ifdef`. That doesn't bother me too much. I do think that going with a regular `Assert` would keep the function fast in release mode.